packages/ui/src/theme/utils.ts

```diff
@@ -48,3 +51,95 @@ interface NameTransformProps {
 export function cssNameTransform({ path = [] }: NameTransformProps): string {
   return `${kebabCase([CSS_VARIABLE_PREFIX, ...path].join(' '))}`;
 }
+
+export type pathObject = {
+  path: string[];
+};
+
+export const STATE_NAMES = [
+  'hover',
+  'focus',
+  'active',
+  'loading',
+  'disabled',
+  'visited',
+];
+
+export function findDeprecatedStateTokens(baseTheme: BaseTheme): pathObject[] {
+  const baseEntry = baseTheme?.tokens || {};
+  return traverse(['', baseEntry], []);
+}
+
+/**
+ * This function will recursively traverse the tokens object from a theme and return a path array to
+ * tokens with the deprecatedStateToken flag. This will later be used to match up against the passed in theme object
+ * and write State token values that are being passed in using the old deprecated state tokens.
+ */
+function traverse(baseEntry, existingPath: string[]): pathObject[] {
+  let path = [...existingPath];
+  const [key, obj] = baseEntry;
+  if (typeof obj === 'object' && !Array.isArray(obj)) {
+    if (key) {
+      path.push(key);
+    }
+    if (obj.value) {
+      if (obj.deprecatedStateToken) {
+        return [
+          {
+            path: path,
+          },
+        ];
+      }
+      return [];
+    } else {
+      const entries = Object.entries(obj);
+      let retValue = [];
+      for (const entry of entries) {
+        let deprecated = traverse(entry, path);
+        if (deprecated.length) {
+          if (Array.isArray(deprecated)) {
+            retValue = [...retValue, ...deprecated];
+          } else {
+            retValue.push(deprecated);
+          }
+        }
+      }
+      if (retValue.length) {
+        return retValue;
+      } else {
+        return [];
+      }
+    }
+  }
+  return [];
+}
```

Comment:
How about simplify it to be something like this? 

```suggestion
function traverseForDeprecated(baseEntry, childrenPath: string[]): pathObject[] {
  let path = [...childrenPath];
  const [key, obj] = baseEntry;
  if (typeof obj === 'object' && !Array.isArray(obj)) {
    if (key) {
      path.push(key);
    }
    if (obj.value) {
      return obj.deprecatedStateToken ? [{ path }] : [];
    } else {
      return Object.entries(obj).flatMap((entry) => traverseForDeprecated(entry, path));
    }
  }
  return [];
}
```

I didn't test it. It's just an idea 😁